yarn-project/pxe/src/pxe_service/error_enriching.ts

```diff
@@ -0,0 +1,81 @@
+import { type SimulationError, isNoirCallStackUnresolved } from '@aztec/circuit-types';
+import { AztecAddress, Fr, FunctionSelector, PUBLIC_DISPATCH_SELECTOR } from '@aztec/circuits.js';
+import { type DebugLogger } from '@aztec/foundation/log';
+import { resolveAssertionMessage, resolveOpcodeLocations } from '@aztec/simulator';
+
+import { type ContractDataOracle, type PxeDatabase } from '../index.js';
+
+/**
+ * Adds contract and function names to a simulation error.
+ * @param err - The error to enrich.
+ */
+export async function enrichSimulationError(err: SimulationError, db: PxeDatabase) {
+  // Maps contract addresses to the set of functions selectors that were in error.
+  // Using strings because map and set don't use .equals()
```

Comment:
I don't understand the comment about `equals`

Reply:
https://stackoverflow.com/a/31953390

Reply:
```suggestion
  // Map and Set do reference equality for their keys instead of value equality, so we store the string
  // representation to get e.g. different contract address objects with the same address value to match.
```